### ISSUE TITLE
Custom object detection by retaining the original classes of yolo

### ISSUE THREAD

**Issue by b4u365** [ultralytics/yolov5#13025]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I have custom data set with Road signs, Traffic barrales which need to be trained and detected **retaining** the earlier classes of yolov5s.pt  like (Cars, pedistrains, etc). once i train with  custom data set , yolo could only detect the trained objects like Traffic barrels but was not detecting the cars and persons. can you help me in resolving this issue were i need to detect both road signs, traffic barrales along with  Cars, pedistians which was there in yolov5 deafult weights .

**Comment by glenn-jocher** [comment]
@b4u365 hello! It sounds like you want to expand your model's capabilities to detect both new and existing classes. To achieve this, you'll need to train on a dataset that includes all desired classes — both the default ones like cars and pedestrians, and your custom classes like road signs and traffic barrels.

Here’s a quick guide:

1. **Dataset Preparation**: Combine your custom dataset with a dataset containing the default YOLO classes. Ensure each class is properly labeled.
   
2. **Training**: When setting up your training configuration in YOLOv5, update the `*.yaml` file to include all classes (default and custom). This means adjusting the number of classes and providing the names of each class.
   
3. **Model Training**: Use the command to start training, ensuring to specify the path to your combined dataset and updated `*.yaml` file.

This approach ensures that the model learns to detect both old and new classes together. Good luck with your training! 😊

**Comment by b4u365** [comment]
Thank you so much, This Helps me alot.

 Can I freeze all the layers except last layers as shown below option A,
will  freezing  transfer the features of weights of yolov5s.pt which was
trained on larger datasets?

or is it advisable to train all the layers (option B)? My main goal is the
weights must be as good as the original yolov5s.pt with additional
capability due to  custom training dataset.


option  A:  python train.py --freeze 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15
16 18 19 21 22
or
option  B:  python train.py.

Can you help me in resolving this?

Thanks and regards,
Bharath.

**Comment by glenn-jocher** [comment]
@b4u365 hi Bharath,

Glad to hear the previous advice was helpful! Regarding your question about freezing layers:

- **Option A (Freezing Layers)**: Freezing most of the layers and only training the last few can indeed help in transferring the learned features from the large dataset used in yolov5s.pt. This approach is faster and consumes less computational resources. However, it might limit the model's ability to adapt to the new classes in your custom dataset.

- **Option B (Training All Layers)**: This option allows the model to better adapt to the new data, potentially improving accuracy for your custom classes at the expense of more extensive computation and training time.

Given your goal of retaining the original quality of yolov5s.pt while adding new capabilities, I'd recommend starting with **Option A** to leverage the pre-trained features and then, if necessary, fine-tuning with **Option B** based on the performance and results you observe.

Best of luck with your training! 😊

**Comment by b4u365** [comment]
Thank you so much for quick response.

I trained using option “A” :

python train.py --freeze 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 18 19 21
22

But the performance for detection of cars has decreased  with option “A”
approach compared to original weights of yolov5 . can you help in getting
the performance similar to original
Weights but want to include the new classes.

Thanks and regards,

Bharath

**Comment by glenn-jocher** [comment]
Hi Bharath,

It sounds like freezing the layers might be too restrictive for adapting to your new classes while maintaining performance on the original ones. I recommend gradually unfreezing some of the earlier layers to allow more flexibility in learning features relevant to both new and existing classes. You can start by unfreezing earlier layers incrementally and observing the impact on performance.

Here's a modified command to unfreeze some earlier layers:
```bash
python train.py --freeze 10 11 12 13 14 15 16 18 19 21 22
```

This approach strikes a balance between leveraging pre-trained features and adapting to new data. Let's see how this adjustment works out! 😊
